feat(Raw Printing): Adding Support for qz-tray

karthikeyan5 opened the pull request:
Screenshots:

- printing by raw commands for the first time
![raw-printing-from-print-view](https://user-images.githubusercontent.com/4847317/56187860-8ad70980-6041-11e9-82b9-fce8b95bb8d3.gif)



- creating a new Raw Commands print format
![raw-command-print-format](https://user-images.githubusercontent.com/4847317/56129366-78a48f00-5f9f-11e9-9808-5b4a102fa875.png)


Documentation PR: https://github.com/frappe/erpnext_com/pull/308

karthikeyan5 committed:
feat(Raw Printing): Adding Support for qz-tray

karthikeyan5 committed:
fix(Raw Printing): fixing code standard

karthikeyan5 committed:
fix(raw printing): codacy issues

karthikeyan5 committed:
fix(raw printing): more codacy fixes

rmehta commented on frappe/public/js/frappe/form/print.js:
```diff
@@ -286,6 +373,76 @@ frappe.ui.form.PrintPreview = Class.extend({
+	},
+	qz_setting_dialog: function () {
+		var me = this;
+		if (localStorage && localStorage.print_format_printer_map)
```
brackets

rmehta commented on frappe/public/build.json:
```diff
@@ -152,7 +152,9 @@
 		"public/js/lib/leaflet/L.Control.Locate.js",
-		"public/js/lib/leaflet/easy-button.js"
+		"public/js/lib/leaflet/easy-button.js",
+		"node_modules/js-sha256/build/sha256.min.js",
```
use `import` ?

karthikeyan5 replied:
used import for sha256

rmehta reviewed (changes requested):
Also add code comments and docs

karthikeyan5 committed:
fix(Raw Printing): fixes as per PR review

> fixing confusing naming in the UI and calling it "Raw Print"
> used import for sha256 library
> changed the Print Preview message for raw print
> same possible codacy fixes

karthikeyan5 commented:
todo:
- [ ] add a print setting called "Enable Raw Printing"
- [ ] Add more comments in the code
- [ ] add documentation about instaling qz-tray right in the UI
- [ ] add documentation about raw commands in the new print format UI
- [ ] make a Documentation PR for "Raw Printing" Feature

karthikeyan5 committed:
fix(raw printing): more codacy fixes

karthikeyan5 committed:
feat(Raw Printing): added "Enable Raw Printing"

> added more comments in code

karthikeyan5 committed:
feat(Raw Printing): added help text in UI

rmehta commented:
@karthikeyan5 lets use "Printer Settings" instead of "Raw Print Settings"

the term "Raw Print" is not very user friendly

rmehta commented on frappe/public/build.json:
```diff
@@ -153,7 +153,8 @@
 		"public/js/lib/leaflet/L.Control.Locate.js",
-		"public/js/lib/leaflet/easy-button.js"
+		"public/js/lib/leaflet/easy-button.js",
+		"node_modules/qz-tray/qz-tray.js"
```
can we load this on demand, with `frappe.require` ?

karthikeyan5 replied:
yes, that makes sense! didn't know `frappe.require` existed...

`qz-tray.js` and `sha256.min.js` are now loaded only when calling the qz.connect function.

rmehta commented on frappe/public/js/frappe/form/print.js:
```diff
@@ -190,6 +212,49 @@ frappe.ui.form.PrintPreview = Class.extend({
+						//Note: need to solve "Error: Cannot parse (FILE)<URL> as a PDF file" to enable qz pdf printing.
+
+						// // use pdf method print method of qz
+						// let printer_map = me.get_mapped_printer()[0]
```
remove commented code

karthikeyan5 replied:
thought it would make it easier to enable PDF printing via QZ Tray after solving the Error... 

I have removed it now.

rmehta commented on frappe/public/js/frappe/form/print.js:
```diff
@@ -326,3 +494,88 @@ frappe.ui.get_print_settings = function (pdf, callback, letter_head) {
+		} else {
+			// try to connect once before firing the mimetype launcher
+			frappe.show_alert({
+				message: __('Attemting Connection to QZ Tray!'),
```
attempting (spelling)

Also no exclamation mark!

karthikeyan5 replied:
oops... ran spell check and corrected a few other spellings as well.

karthikeyan5 committed:
fix(Raw Printing): fixes as per PR review 2

> qz-tray.js and sha256.min.js are loaded only on demand
> renamed 'raw print settings' button to 'Printer Settings'
> removed commented code
> fixed spelling mistakes

TurkerTunali commented:
> 
> 
> @karthikeyan5 lets use "Printer Settings" instead of "Raw Print Settings"
> 
> the term "Raw Print" is not very user friendly

"Direct Print" may be more suitable. I've looked other ERP systems but they just name it as "Print". There are no more options.

karthikeyan5 commented:
> "Direct Print" may be more suitable.

@TurkerTunali "Direct Print" is a whole other thing according to most [sources](https://www.google.co.in/search?q=direct+print). It is related to sending a print directly from a device(like smartphones, desktops) without the need for a print server. Whereas, Raw Printing is related to sending native print commands(AKA Raw Commands) to a printer. 
Also, the end user just sees a "print" button and not a "Raw Print" Button.

Please take a look at the [ERPNext Raw Printing Documentation](https://erpnext.com/docs/user/manual/en/setting-up/print/raw-printing) for more clarity on Raw Printing.

dexterhussain commented:
Can someone please tell me how i can use this?

karthikeyan5 commented:
> Can someone please tell me how i can use this?

@dexterhussain, have you gone through the documentation? https://erpnext.com/docs/user/manual/en/setting-up/print/raw-printing

Please ask specific questions after you have tried the steps in the documentation....

dexterhussain commented:
Yes what i cannot understand is how serial number will be made barcode?

On Mon, Mar 2, 2020 at 1:44 PM Karthikeyan S <notifications@github.com>
wrote:

> Can someone please tell me how i can use this?
>
> @dexterhussain <https://github.com/dexterhussain>, have you gone through
> the documentation?
> https://erpnext.com/docs/user/manual/en/setting-up/print/raw-printing
>
> Please ask specific questions after you have tried the steps in the
> documentation....
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/frappe/frappe/pull/7088?email_source=notifications&email_token=AB6RS36WPADRH6FBDS4FCYTRFNWVHA5CNFSM4G7HBP7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENONMBY#issuecomment-593286663>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB6RS35V5TSLRQ2AXWUFQ43RFNWVHANCNFSM4G7HBP7A>
> .
>


-- 

Regards,



Muzzammil Hussain

Principal Software Engineer/ Team Lead

Tower Technologies (Pvt.) Ltd.

Office No. 04, 7th Floor,

Arfa Software Technology Park,

346-B, Ferozepur Road,

Lahore, Pakistan

Office: +92-42-35972042 & 43

Cell: +92-321-8585322

Website: www.thetowertech.com

dexterhussain commented:
To elaborate my issue
can someone please help me in this?
https://discuss.erpnext.com/t/multi-serial-numbers-as-print/58815

tresf commented on yarn.lock:
```diff
@@ -2452,6 +2452,11 @@ js-base64@^2.1.8, js-base64@^2.1.9:
+  resolved "https://registry.yarnpkg.com/js-sha256/-/js-sha256-0.9.0.tgz#0b89ac166583e91ef9123644bd3c5334ce9d0966"
+  integrity sha512-sga3MHh9sgQN2+pJ9VYZ+1LPwXOxuBJBA5nrR5/ofPfuiJBE2hnjsaN8se8JznOmGLN2p49Pe5U/ttafcs/apA==
+
 js-yaml@^3.12.0, js-yaml@^3.9.0:
```
FYI, not sure where else to put this, but we've bundled the sha-256 library starting with 2.1.1 per https://github.com/qzind/tray/pull/620.  You're on 2.0.8, so you'll still need this, but just an FYI to the maintainer of the integration. 🍻

tresf commented on frappe/public/js/frappe/form/print.js:
```diff
@@ -326,3 +469,119 @@ frappe.ui.get_print_settings = function (pdf, callback, letter_head) {
+			frappe.require(qz_required_assets,() => {
+				qz.api.setPromiseType(function promise(resolver) {
+					return new Promise(resolver);
+				});
```
Just an FYI, we do this internally for you starting with 2.1.1 (as well as the sha256 stuff in the lines following). 🍻
